### PR TITLE
FEI-3758.3: Get most of the jest tests working

### PR DESCRIPTION
**Pull request by kevinbarabash**
## Summary:
While most of the tests are work there are still a large number of failures that need to be resolved.  The plan is to land this in the current state so that we can start parallelizing the work of resolving remaining issues.

Issue: FEI-3758

## Test plan:
- yarn test, see that most tests are passing

```
Test Suites: 14 failed, 2 skipped, 57 passed, 71 of 73 total
Tests:       97 failed, 18 skipped, 845 passed, 960 total
Snapshots:   7 failed, 2 obsolete, 126 passed, 133 total
Time:        77.591 s, estimated 97 s
```

### CODE REVIEW

File: config/test/test-setup.js [hunk 1]
```diff
@@ -1 +1,192 @@
 // No setup to perform
```
Comment (jeremywiebe) [inline config/test/test-setup.js:1]: I think this file says otherwise. 😂  

✂️

Comment (jeremywiebe) [inline config/test/test-setup.js:1]: I think this file says otherwise. 😂  

✂️ 

```suggestion
```

File: config/test/test-setup.js [hunk 2]
```diff
@@ -1 +1,192 @@
...
+    };
+}
+
+// javascript/shared-package/api-action-results.js sets an ajaxSetup method that
+// attaches an X-KA-FKey header to all AJAX requests based on the "fkey" cookie
+// value. If it doesn't find the cookie, it blows up, so make sure it's always
+// there.
+if (typeof document !== "undefined") {
+    document.cookie = "fkey=FAKEXSRFTOKEN";
+}
```
Comment (jeremywiebe) [inline config/test/test-setup.js:69]: We probably (hopefully) don't need this.

File: config/test/test-setup.js [hunk 3]
```diff
@@ -1 +1,192 @@
...
+// "accessing an undefined variable" error in load-mathjax.js), but it doesn't
+// have to be anything meaningful. If we add tests that require this, then we
+// can copy the value from webpack.common.js.
+globalThis.WEBPACK_DEFINE_MATHJAX_URL =
+    "we-dont-currently-test-mathjax-i-guess";
+
+// We don't run webpack for our tests, but some code expects some webpack
+// utilities. Rather than have prod code always affirming what we know will
+// be there, let's just sub things in.
+globalThis.__webpack_public_path__ = "/genwebpack/services/static/";
```
Comment (jeremywiebe) [inline config/test/test-setup.js:83]: Also suspect we don't need this.

File: config/test/test-setup.js [hunk 4]
```diff
@@ -1 +1,192 @@
...
+};
+const registerConsoleWarnsOrErrorCall = () => {
+    consoleWarnsAndErrors = true;
+};
+const reportUnhandledConsoleWarnAndErrors = () => {
+    if (consoleWarnsAndErrors) {
+        resetConsoleWarnsAndErrors();
+        throw new Error(
+            "Unhandled console warnings or errors.\n" +
+                "Use perseus-all-package/logging/log.js to log, and make " +
```
Comment (jeremywiebe) [inline config/test/test-setup.js:104]: Update path?

File: config/test/test-setup.js [hunk 5]
```diff
@@ -1 +1,192 @@
...
+    // Make sure that any console errors or warnings trigger a test failure.
+    // Overwhelmingly these failures are due to real errors that are being
+    // hidden, so we should be attempting to fix them.
+    reportUnhandledConsoleWarnAndErrors();
+});
+
+afterAll(() => {
+    // We track to see if there were any console warnings or errors that came
+    // up *after* all the tests finished. This appears to happen for GraphQL
+    // failures (they are being executed async)
```
Comment (jeremywiebe) [inline config/test/test-setup.js:190]:
```suggestion
    // We track to see if there were any console warnings or errors that came
    // up *after* all the tests finished. 
```

File: config/test/test.config.js [hunk 6]
```diff
@@ -12,15 +12,27 @@ module.exports = {
     },
     restoreMocks: true,
     resetMocks: true,
-    testEnvironment: "jest-environment-node",
-    testMatch: ["<rootDir>/**/*.test.js"],
+    testEnvironment: "jsdom",
+    testMatch: [
+        "<rootDir>/**/*.test.js",
+        "<rootDir>/**/*_test.js",
```
Comment (jeremywiebe) [inline config/test/test.config.js:18]: I suspect we don't need this twice?

```suggestion
```

File: config/test/test.config.js [hunk 7]
```diff
@@ -12,15 +12,27 @@ module.exports = {
...
+        "<rootDir>/**/*_test.jsx",
+    ],
+    timers: "fake",
```
Comment (jeremywiebe) [inline config/test/test.config.js:21]: Given this, do we need to do the `jest.useFakeTimers()` in `config/test/test-setup.js`?

File: config/test/test.config.js [hunk 8]
```diff
@@ -12,15 +12,27 @@ module.exports = {
...
         "<rootDir>/config/test/test-setup.js",
+        "<rootDir>/config/test/custom-matchers.js",
     ],
     moduleNameMapper: {
         "^@khanacademy/perseus(.*)$":
             "<rootDir>/packages/perseus$1/src/index.js",
+        // TODO: generate this mapping automatically
+        hubble: "<rootDir>/vendor/hubble/hubble.js",
+        jsdiff: "<rootDir>/vendor/jsdiff/jsdiff.js",
+        raphael: "<rootDir>/vendor/raphael/raphael.js",
```
Comment (jeremywiebe) [inline config/test/test.config.js:33]: I've wanted to do a map => object so many times.  I think it'd even be useful within perseus as we start to replace underscore. 

Something like this would be useful here:
```js
...objMap(["nibble", "jsdiff", "raphael"], key => {
    return [
        key, 
        `<rootDir>/vendor/${key}/${key}.js`,
    ];
}

```

No action required.

File: flow-typed/npm/jest_v27.x.x.js [hunk 9]
```diff
@@ -1,6 +1,35 @@
 // flow-typed signature: 5ddcf688200e3506308fdcfa78ca48d9
 // flow-typed version: 644a595e77/jest_v27.x.x/flow_>=v0.134.x
 
+/**
```
Comment (jeremywiebe) [inline flow-typed/npm/jest_v27.x.x.js:4]: Should we do like we do in webapp and the `_CONTAINS_KA_CUSTOMIZATIONS` bit to the filename also?

File: packages/perseus-editor/src/widgets/label-image/select-image.jsx [hunk 10]
```diff
@@ -3,15 +3,14 @@
  * Controlled component for selecting the question image, for answer labeling.
  */
 
+import {globalStyles} from "@khanacademy/perseus";
```
Comment (jeremywiebe) [inline packages/perseus-editor/src/widgets/label-image/select-image.jsx:6]: Could you make a ticket to clean up (ie. delete) `globalStyles` and move to use WB stuff?  I'm not sure how big a task that is. I know in some cases a component uses a single color out of the styles, but some might be more complex.

File: packages/perseus/src/interactive2/movable-line-options.js [hunk 11]
```diff
@@ -3,8 +3,7 @@
 /**
  * A library of options to pass to add/draw/remove/constraints
  */
-import {vector as kvector} from "kmath";
-import {point as kpoint} from "kmath";
+import {vector as kvector, point as kpoint} from "kmath";
```
Comment (jeremywiebe) [inline packages/perseus/src/interactive2/movable-line-options.js:6]: Thanks for collapsing these!  Do you have tooling to help or just "brute force" 😁

File: packages/perseus/src/styles/global-constants.js [hunk 12]
```diff
@@ -0,0 +1,350 @@
+// @flow
```
Comment (jeremywiebe) [inline packages/perseus/src/styles/global-constants.js:1]: I hope we can make this really short-lived. :)

File: packages/perseus/src/util/interactive.js [hunk 13]
```diff
@@ -10,15 +10,14 @@
...
-import {line as kline} from "kmath";
+import {vector as kvector, point as kpoint, line as kline} from "kmath";
 import Raphael from "raphael";
 import _ from "underscore";
 
 // NOTE(jeresig): We include a special copy of jQuery vmouse so
 // that Perseus will continue to work in mobile native devices
 // (this should have no impact in the browser)
+// eslint-disable-next-line import/no-unassigned-import
 import "../jquery.mobile.vmouse.js";
```
Comment (jeremywiebe) [inline packages/perseus/src/util/interactive.js:21]: Oh. This looks like breaking convention for importing a jQuery plugin through our jQuery bundle.  Do we really have this file in two places in webapp?

File: packages/perseus/src/widgets/__tests__/grapher.cypress.jsx [hunk 14]
```diff
@@ -27,7 +27,7 @@ const LINES = "[data-interactive-kind-for-testing=movable-line] > svg > path";
 describe("Grapher widget", () => {
     beforeEach(() => {
         Perseus.init({skipMathJax: true});
-        Perseus.Dependencies.setDependencies(webappDependencies);
+        Perseus.Dependencies.setDependencies(testDependencies);
```
Comment (jeremywiebe) [inline packages/perseus/src/widgets/__tests__/grapher.cypress.jsx:30]: I suspect this might not work (pretty sure I had to use webappDependencies when I was getting Cypress working after we introduced dependencies). But we can fix this when we get to the Cypress setup ticket.

File: testing/README.md [hunk 15]
```diff
@@ -1,4 +1,4 @@
-# perseus-testing
+# ../../testing
```
Comment (jeremywiebe) [inline testing/README.md:1]: Maybe just `# Testing`? 😄

**Review by jeremywiebe — APPROVED** [review]
Mostly little cleanup comments, but after that, LGTM.

File: config/test/test.config.js (hunk 6 above)
Comment (kevinbarabash) [reply config/test/test.config.js:18]: These are slightly different.  I'll remove the `*.test.js` for now, but it would be nice to transition to that format since VSCode recognizes `.test.js` files as tests.

File: packages/perseus/src/util/interactive.js (hunk 13 above)
Comment (kevinbarabash) [reply packages/perseus/src/util/interactive.js:21]: I sounds like this copy of vmouse.js is different from what's included in our jquery bundle in webapp.  We should probably verify that and if it isn't we can get rid of it.  Since this is currently the status quo I'm going to leave it for now.

File: packages/perseus/src/widgets/__tests__/grapher.cypress.jsx (hunk 14 above)
Comment (kevinbarabash) [reply packages/perseus/src/widgets/__tests__/grapher.cypress.jsx:30]: If that's the case then we need to have a version of `testDependencies` that does more robust mocking.

File: testing/README.md (hunk 15 above)
Comment (kevinbarabash) [reply testing/README.md:1]: TBH, I think we can nuke this file since we're consolidating the various coverage reports into a single repo.
